lib/proxy/peer/quic/quic.go

```diff
@@ -0,0 +1,189 @@
+are not relevant; data sent by the server as a response to the client is either
+using 0-RTT keys or is sent after the handshake is completed.
+*/
+package quic
+
+import (
+	"encoding/binary"
+	"errors"
+	"math"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/quic-go/quic-go"
+	"github.com/quic-go/quic-go/quicvarint"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	// quicMaxIdleTimeout is the arbitrary timeout after which a QUIC connection
+	// that hasn't received data is presumed to be lost to the aether.
+	quicMaxIdleTimeout = 30 * time.Second
+	// quicKeepAlivePeriod is the interval of QUIC keepalive packets sent if the
+	// connection is otherwise idle.
+	quicKeepAlivePeriod = 5 * time.Second
+
+	quicMaxReceiveWindow   = quicvarint.Max
+	quicMaxIncomingStreams = 1 << 60 // maximum allowed value as per the quic-go docs
+
+	// quicNextProto is the ALPN indicator for the current version of the QUIC
+	// proxy peering protocol.
+	quicNextProto = "teleport-peer-v1alpha"
+
+	// quicMaxMessageSize is the maximum accepted size (in protobuf binary
+	// format) for the request and response messages exchanged as part of the
+	// dialing.
+	quicMaxMessageSize = 128 * 1024
+
+	// quicTimestampGraceWindow is the maximum time difference between local
+	// time and reported time in a 0-RTT request. Clients should not keep trying
+	// to use a request after this much time has passed.
+	quicTimestampGraceWindow = time.Minute
+	// quicNoncePersistence is the shortest time for which a nonce will be kept
+	// in memory to prevent 0-RTT replay attacks. Should be significantly longer
+	// than [quicTimestampGraceWindow]. In the current implementation, nonces
+	// can be kept for at most twice this value.
+	quicNoncePersistence = 5 * time.Minute
+
+	quicDialTimeout          = 30 * time.Second
+	quicRequestTimeout       = 10 * time.Second
+	quicErrorResponseTimeout = 10 * time.Second
+)
+
+// marshalSized returns the wire encoding of the given [proto.Message] prefixed
+// by its length encoded as a little endian 32-bit integer.
+func marshalSized(m proto.Message) ([]byte, error) {
+	buf := make([]byte, 4, 4+proto.MarshalOptions{}.Size(m))
+	buf, err := proto.MarshalOptions{UseCachedSize: true}.MarshalAppend(buf, m)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(buf)-4 > math.MaxUint32 {
+		return nil, trace.LimitExceeded("oversized message")
+	}
+	binary.LittleEndian.PutUint32(buf, uint32(len(buf)-4))
+	return buf, nil
+}
+
+// ignoreCodeZero returns nil if the given error is a [quic.StreamError] or
+// [quic.ApplicationError] with an error code of 0; the error is returned
+// [trace.Wrap]ped otherwise.
+func ignoreCodeZero(err error) error {
+	if streamErr := (*quic.StreamError)(nil); errors.As(err, &streamErr) && streamErr.ErrorCode == 0 {
+		return nil
+	}
+	if appErr := (*quic.ApplicationError)(nil); errors.As(err, &appErr) && appErr.ErrorCode == 0 {
+		return nil
+	}
+	return trace.Wrap(err)
```

Comment:
I don't think the trace.Wrap belongs here, let the caller wrap if they want to. It's documented, which is good, but it's still surprising.